Source/WebCore/rendering/style/StyleGradientImage.cpp

```diff
@@ -27,74 +27,1101 @@
+            result->setFirstX(data.firstX.copyRef());
+            result->setFirstY(data.firstY.copyRef());
+            result->setSecondX(data.secondX.copyRef());
+            result->setSecondY(data.secondY.copyRef());
+            result->setFirstRadius(data.firstRadius.copyRef());
+            result->setSecondRadius(data.secondRadius.copyRef());
+            result->setShape(data.shape.copyRef());
+            result->setSizingBehavior(data.sizingBehavior.copyRef());
+            result->setEndHorizontalSize(data.endHorizontalSize.copyRef());
+            result->setEndVerticalSize(data.endVerticalSize.copyRef());
+            return result;
+        },
+        [&] (const ConicData& data) -> Ref<CSSValue> {
+            auto result = CSSConicGradientValue::create(m_repeat, m_colorInterpolationMethod, WTFMove(cssStopList));
+            result->setFirstX(data.firstX.copyRef());
+            result->setFirstY(data.firstY.copyRef());
+            result->setSecondX(data.secondX.copyRef());
+            result->setSecondY(data.secondY.copyRef());
+            result->setAngle(data.angle.copyRef());
+            return result;
+        }
+    );
 }
 
 bool StyleGradientImage::isPending() const
 {
-    return m_imageGeneratorValue->isPending();
+    return false;
 }
 
-void StyleGradientImage::load(CachedResourceLoader& loader, const ResourceLoaderOptions& options)
+void StyleGradientImage::load(CachedResourceLoader&, const ResourceLoaderOptions&)
 {
-    m_imageGeneratorValue->loadSubimages(loader, options);
 }
 
 RefPtr<Image> StyleGradientImage::image(const RenderElement* renderer, const FloatSize& size) const
 {
-    return renderer ? m_imageGeneratorValue->image(const_cast<RenderElement&>(*renderer), size) : &Image::nullImage();
+    if (!renderer)
+        return &Image::nullImage();
+
+    if (size.isEmpty())
+        return nullptr;
+
+    bool cacheable = isCacheable() && !renderer->style().hasAppleColorFilter();
+    if (cacheable) {
+        if (!clients().contains(const_cast<RenderElement*>(renderer)))
+            return nullptr;
+        if (auto* result = const_cast<StyleGradientImage&>(*this).cachedImageForSize(size))
+            return result;
+    }
+
+    auto gradient = WTF::switchOn(m_data,
+        [&] (auto& data) -> Ref<Gradient> {
+            return createGradient(data, *renderer, size);
+        }
+    );
+
+    auto newImage = GradientImage::create(WTFMove(gradient), size);
+    if (cacheable)
+        const_cast<StyleGradientImage&>(*this).saveCachedImageForSize(size, newImage);
+    return newImage;
 }
 
 bool StyleGradientImage::knownToBeOpaque(const RenderElement& renderer) const
 {
-    return m_imageGeneratorValue->knownToBeOpaque(renderer);
+    bool hasColorFilter = renderer.style().hasAppleColorFilter();
+    for (auto& stop : m_stops) {
+        Color color = stop.resolvedColor;
```

Comment:
A shame to make a copy when the color is extended and `hasColorFilter` is false.